### DISCUSSION TITLE
Lesson 11: Repository Update

### DISCUSSION THREAD

**Discussion by Niferu** [smartcontractkit/full-blockchain-solidity-course-py#1700] [Show and tell]
Hello,

I have found small issue with repo code in Lesson 11. As I'm new to programming and GitHub, let me how can i rise such discoveries to for example fix repo code a bit.

Problem i have found is located in `scripts.advanced_collectible.create_metadata.py`:
As we are using `token_id` to provide name for our `.json` ipfs link we end up using incorrect number per each `breed` as it is just counter for our loop.

Below is the .json link generated for ST_BERNARD with current repo code:
![image](https://user-images.githubusercontent.com/110537615/190192728-cb948b72-94a9-4ab8-8b8c-6fc6d3fdf894.png)

As from mapping it is nr. 2 not 0, link to it is different than in `scripts.advanced_collectible.set_token_uri.py` mapping.

In order to generate similar link from `scripts.advanced_collectible.set_token_uri.py` from `dog_metadata_dic`. We should change our code in `create_metadata.py`. 

I have prepared corrected code below:
```py
from scripts.helpful_scripts import INVERSE_BREED_MAPPING, get_breed

# from helpful_scripts: INVERSE_BREED_MAPPING = {"PUG": 0, "SHIBA_INU": 1, "ST_BERNARD": 2}


def main():
    advanced_collectible = AdvancedCollectible[-1]
    number_of_advanced_collectibles = advanced_collectible.tokenCounter()
    print(f'You Have Created {number_of_advanced_collectibles} Collectibles!')
    for token_id in range(number_of_advanced_collectibles):
        breed = get_breed(advanced_collectible.tokenIdToBreed(token_id))
        dog_index = INVERSE_BREED_MAPPING[breed]
        metadata_file_name = (f'./metadata/{network.show_active()}/{dog_index}-{breed}.json')
        ...
```
Results of above code:
![image](https://user-images.githubusercontent.com/110537615/190202474-a9c3dc08-ffd8-447a-8eb0-530a53dedba6.png)

I couldn't get same .json link as it is in mapping, but i believe this is because my .json link shows everything in 1 line as below:
![image](https://user-images.githubusercontent.com/110537615/190202142-8c9cae66-3d72-469b-b6aa-45b7fc09ba9d.png)

While repo mapping link shows it .json formatted as below.
![image](https://user-images.githubusercontent.com/110537615/190201884-5dd7393f-d974-4bfe-8152-00ced270d57a.png)

 How can this be fixed so our generated link shows also formatted data?

**Comment by cromewar** [comment]
Hello @Neftyr 
For someone who is new to coding you are really good at this, congrats.
If I understood correctly yo are asking about the mismatch on the indexes when running the code right?  because the breed number (id) does not correspond the Json number the file is generated.

Well the easiest way to avoid this is just to delete the build folder, as the latest deployment will be erased and the new contract will start at token_id 0 so this will match the json files. (also delete the generated json files before running again).

However you approach is awesome also, congrats again.

Regarding your question, well it depends on the browser you are using as most of them will just read ipfs/json files as plain text without any type of formatting, to have the generated link formatted on your browser you will need an extension, I highly recommend to try out:

The official [IPFS](https://chrome.google.com/webstore/detail/ipfs-companion/nibjojkomfdiaoajekhjakgkdhaomnch?hl=es) extension.
A [Json formatter](https://chrome.google.com/webstore/detail/json-formatter/bcjindcccaagfpapjjmafapmmgkkhgoa?hl=es) extension.

I hope this info might help.
Cheers.

**Reply by Niferu** [reply]
Hi @cromewar 
Thanks a lot for kind words, that makes me even more motivated!

Regarding this case, you are totally correct I meant the mismatch on the indexes when running the code and yes breed number (id) does not correspond the Json number the file is generated. I have tried removing build folder and i have also removed all generated .json files but as I expected index is still the same as it is loop counter given for name not index of breed, so it doesnt matter, as If we deploy 1 contract, counter will point always index 0 and our file will be always named 0-"breed_name" no matter what breed has been generated at random. In my opinion if my 1st contract deployed generated for example SHIBA_INU, .json file should be named 1-shiba_inu and not 0-shiba_inu as it is for json files in repo, if contract generated ST_BERNARD my name should be 2-ST_BERNARD etc.

Regarding second part, i have made both screens from same browser after opening link generated by me and link added to repo, so I doubt it is because of browser as formatting was correct for repo link and incorrect for mine link.

**Reply by cromewar** [reply]
Take in mind on coding in general we always start counting in "0" using 0 as the first index is actually the correct approach.
Which browser are you using?
Firefox has native support for json formatted pages.
I recommend using one of those extensions on Brave browser.

**Reply by Niferu** [reply]
Hello @cromewar,

Yes thank you I know iterating should start from 0 in coding, but I suppose I have misinterpreted intention of this function and I just thought we should always match name to breed, but obviously we should just count NFT deployment's and match them with names, so I think I got it now, the `.json` files in repository has been just made in order and that is why their file names matches breed number, so PUG "0", SHIBA "1" and BERNARD "2" as PUG has been generated 1st etc. Sorry for misunderstanding I was hoping to improve repo code here haha, maybe one day it will happen.

Regarding browser and format thing, I'm using `Chrome`, I have tested `Firefox` for file generated by myself and indeed it is formatted there. I was just curious why `repo .json link` is formatted even on my Chrome without any extensions while `mine link` is not. If you have any idea let me know if not I will just get that extension, so all .json links in Chrome can be read same way. Thanks a lot for your help!

**Reply by cromewar** [reply]
Actually if you feel that you can make any improvement on the repo feel free to create a Pull Request with the changes, and Patrick or myself  would give them a check and accept the changes if they are good ;).

Regarding to your second question, well I have no clue. I wonder myself why sometimes some json files get automatically formatted on the browser and some of them not. I'll try to do more research on this though. 

Cheers.
